pkg/chunk/purger/purger.go

```diff
@@ -0,0 +1,578 @@
+}
+
+type workerJob struct {
+	planNo          int
+	userID          string
+	deleteRequestID string
+	logger          log.Logger
+}
+
+type workerJobExecutionStatus struct {
+	workerJob
+	err error
+}
+
+// DataPurger does the purging of data which is requested to be deleted
+type DataPurger struct {
+	cfg           DataPurgerConfig
+	deleteStore   *chunk.DeleteStore
+	chunkStore    chunk.Store
+	storageClient chunk.ObjectClient
+
+	executePlansChan             chan deleteRequestWithLogger
+	workerJobChan                chan workerJob
+	workerJobExecutionStatusChan chan workerJobExecutionStatus
+
+	// we would only allow processing of singe delete request at a time since delete requests touching same chunks could change the chunk IDs of partially deleted chunks
+	// and break the purge plan for other requests
+	inProcessRequestIDs    map[string]string
+	inProcessRequestIDsMtx sync.RWMutex
+
+	quit chan struct{}
+	wg   sync.WaitGroup
+}
+
+// NewDataPurger creates a new DataPurger
+func NewDataPurger(cfg DataPurgerConfig, deleteStore *chunk.DeleteStore, chunkStore chunk.Store, storageClient chunk.ObjectClient) (*DataPurger, error) {
+	dataPurger := DataPurger{
+		cfg:                          cfg,
+		deleteStore:                  deleteStore,
+		chunkStore:                   chunkStore,
+		storageClient:                storageClient,
+		inProcessRequestIDs:          map[string]string{},
+		executePlansChan:             make(chan deleteRequestWithLogger, 50),
+		workerJobChan:                make(chan workerJob, 50),
+		workerJobExecutionStatusChan: make(chan workerJobExecutionStatus, 50),
+		quit:                         make(chan struct{}),
+	}
+
+	dataPurger.runWorkers()
+	go dataPurger.jobScheduler()
+
+	err := dataPurger.loadInprocessDeleteRequests()
+	if err != nil {
+		return nil, err
+	}
+
+	return &dataPurger, nil
+}
+
+// Run starts workers, job scheduler and a loop for feeding delete requests to them for processing
+func (dp *DataPurger) Run() {
+	dp.loop()
+}
+
+// Stop stops all background workers/loops
+func (dp *DataPurger) Stop() {
+	close(dp.quit)
+	dp.wg.Wait()
+}
+
+func (dp *DataPurger) loop() {
+	dp.wg.Add(1)
+	defer dp.wg.Done()
+
+	pullDeleteRequestsToPlanDeletesTicker := time.NewTicker(time.Hour)
+	defer pullDeleteRequestsToPlanDeletesTicker.Stop()
+
+	for {
+		select {
+		case <-pullDeleteRequestsToPlanDeletesTicker.C:
+			err := dp.pullDeleteRequestsToPlanDeletes()
+			if err != nil {
+				level.Error(util.Logger).Log("msg", "error pulling delete requests for building plans", "err", err)
+			}
+		case <-dp.quit:
+			return
+		}
+	}
+}
+
+// we send all the delete plans to workerJobChan and then start checking for status on workerJobExecutionStatusChan
+func (dp *DataPurger) jobScheduler() {
+
+	pendingPlansCount := make(map[string]int)
+	pendingPlansCountMtx := sync.Mutex{}
+
+	dp.wg.Add(1)
+
+	go func() {
+
+		for {
+			select {
+			case jobExecutionStatus := <-dp.workerJobExecutionStatusChan:
```

Comment:
Can you shorten this variable to either `job`/`jobStatus`? In general descriptive explicitly descriptive variable names are good. However, in this case the lines are running long and indented. Also the channel name does a lot of the work in describing what the variable is here.